File: android/app/src/main/kotlin/dev/odeon/android/dados/EnderecoDoServidor.kt
package dev.odeon.android.dados

/// Normalização do endereço do servidor.
///
/// ## É um porte, e a origem está dita
///
/// Isto é o `ServerUrl.kt` de `clients/shared`, traduzido pro idioma deste
/// projeto. **É a terceira cópia do mesmo conhecimento** — a primeira é a do
/// KMP, a segunda é a que a web faz à mão em `web/src/api.ts`.
///
/// A espec assumiu esse custo quando decidiu o app nativo (§1), e é por isso que
/// ele mora dentro do `odeon-client` e não em repositório próprio: num
/// repositório só, um `grep` alcança as três.
///
/// ## O que ele resolve
///
/// O que a pessoa digita é `rog` ou `10.0.2.2`, não `https://rog:8443/`. Exigir
/// o formato exato é o tipo de atrito que faz alguém achar que o servidor está
/// fora.
///
/// E "qual esquema?" não deveria ser pergunta: tenta **https primeiro** e cai
/// pra http se ninguém responder. Quem sabe qual dos dois está ligado é o
/// servidor, não quem digita.
object EnderecoDoServidor {

    /// As portas padrão do Odeon quando não vem uma.
    const val PORTA_HTTPS = 8443
    const val PORTA_HTTP = 8080

    /// Limpa e completa o que foi digitado. `null` quando não sobra host.
    ///
    /// Preserva o esquema se veio um; não inventa porta se veio uma.
    fun normalizar(digitado: String): String? {
        val limpo = digitado.trim()
        if (limpo.isEmpty()) return null

        // O esquema sai ANTES de qualquer limpeza de barra — senão o `//` dele
        // é confundido com barra sobrando e `https://` vira `https:`.
        val esquema = when {
            limpo.startsWith("https://") -> "https://"
            limpo.startsWith("http://") -> "http://"
            else -> null
        }
        val resto = esquema?.let { limpo.removePrefix(it) } ?: limpo

        // Só o host importa; caminho digitado por engano é descartado.
        val hostPorta = resto.substringBefore('/').trim()
        if (hostPorta.isEmpty()) return null

        // Um host precisa ter alguma coisa além de pontuação.
        if (hostPorta.none { it.isLetterOrDigit() }) return null

        return (esquema ?: "") + hostPorta
    }

    /// Endereços a tentar, em ordem.
    ///
    /// Com esquema explícito, respeita a escolha e não tenta o outro — se a
    /// pessoa escreveu `http://`, tentar https por baixo seria surpresa.
    fun candidatos(digitado: String): List<String> {
        val normal = normalizar(digitado) ?: return emptyList()

        if (normal.startsWith("http://") || normal.startsWith("https://")) {
            return listOf(normal)
        }

        val temPorta = normal.substringAfterLast(':', "").toIntOrNull() != null

        return if (temPorta) {
            // Porta explícita sem esquema: pode ser qualquer um dos dois.
            listOf("https://$normal", "http://$normal")
        } else {
            listOf(
                "https://$normal:$PORTA_HTTPS",
                "http://$normal:$PORTA_HTTP",
            )
        }
    }

    fun eSeguro(url: String): Boolean = url.startsWith("https://")
}

File: ios/Odeon/Dados/EnderecoDoServidor.swift
import Foundation

/// Normalização do endereço do servidor.
///
/// ## ⚠️ É a **quarta** cópia deste conhecimento, e o número importa
///
/// A primeira é o `ServerUrl.kt` do KMP em `clients/shared`; a segunda é a que a
/// web faz à mão em `web/src/api.ts`; a terceira é o `EnderecoDoServidor.kt` do
/// Android. Esta é a quarta.
///
/// A espec assumiu esse custo quando escolheu clientes nativos, e é por isso que
/// todos moram no mesmo repositório: **num repositório só, um `grep` alcança as
/// quatro**. Quem mudar a regra de porta padrão aqui tem que mudar nos outros
/// três — e o `PLANO.md` §1 registra isso como dívida assumida de olhos abertos.
///
/// ## O que ele resolve
///
/// O que a pessoa digita é `rog` ou `100.77.253.18`, não `https://rog:8443/`.
/// Exigir o formato exato é o tipo de atrito que faz alguém achar que o servidor
/// está fora do ar.
///
/// E «qual esquema?» não deveria ser pergunta: tenta **https primeiro** e cai pra
/// http se ninguém responder. Quem sabe qual dos dois está ligado é o servidor,
/// não quem digita.
enum EnderecoDoServidor {
    /// As portas padrão do Odeon quando não vem uma.
    static let portaHTTPS = 8443
    static let portaHTTP = 8080

    /// Limpa e completa o que foi digitado. `nil` quando não sobra host.
    ///
    /// Preserva o esquema se veio um; não inventa porta se veio uma.
    static func normalizar(_ digitado: String) -> String? {
        let limpo = digitado.trimmingCharacters(in: .whitespacesAndNewlines)
        if limpo.isEmpty { return nil }

        // ⚠️ O esquema sai ANTES de qualquer limpeza de barra — senão o `//` dele
        // é confundido com barra sobrando e `https://` vira `https:`.
        let esquema: String? =
            if limpo.hasPrefix("https://") { "https://" }
            else if limpo.hasPrefix("http://") { "http://" }
            else { nil }

        let resto = esquema.map { String(limpo.dropFirst($0.count)) } ?? limpo

        // Só o host importa; caminho digitado por engano é descartado.
        let hostPorta = resto.prefix(while: { $0 != "/" })
            .trimmingCharacters(in: .whitespaces)
        if hostPorta.isEmpty { return nil }

        // Um host precisa ter alguma coisa além de pontuação.
        if !hostPorta.contains(where: { $0.isLetter || $0.isNumber }) { return nil }

        return (esquema ?? "") + hostPorta
    }

    /// Endereços a tentar, em ordem.
    ///
    /// ⚠️ Com esquema explícito, respeita a escolha e **não** tenta o outro — se
    /// a pessoa escreveu `http://`, tentar https por baixo seria surpresa.
    static func candidatos(_ digitado: String) -> [String] {
        guard let normal = normalizar(digitado) else { return [] }

        if normal.hasPrefix("http://") || normal.hasPrefix("https://") {
            return [normal]
        }

        let depoisDosDoisPontos = normal.split(separator: ":").last.map(String.init) ?? ""
        let temPorta = Int(depoisDosDoisPontos) != nil

        if temPorta {
            // Porta explícita sem esquema: pode ser qualquer um dos dois.
            return ["https://\(normal)", "http://\(normal)"]
        }
        return [
            "https://\(normal):\(portaHTTPS)",
            "http://\(normal):\(portaHTTP)",
        ]
    }

    static func eSeguro(_ url: String) -> Bool { url.hasPrefix("https://") }
}
